File: src/natives/csharp/source/IPC/ToolRequest.cs
using Newtonsoft.Json.Linq;
using Repeat.ipc;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Repeat.IPC {
    public class ToolRequest : RequestGenerator {
        public ToolRequest(RepeatClient client) : base(client) {
            this.Type = "action";
            this.Device = "tool";
        }

        public string GetClipboard() {
            Action = "get_clipboard";
            ClearParams();
            JToken result = SendRequest();
            if (result == null) {
                Console.WriteLine("Nope nope nope");
                return null;
            }
            return result.Value<string>();
        }

        public bool SetClipboard(string data) {
            Action = "set_clipboard";
            ParamInt.Clear();
            ParamStrings.Clear();
            ParamStrings.Add(data);
            return SendRequest() == null ? false : true;
        }

        public string Execute(string cmd, string cwd = null) {
            Action = "execute";
            ClearParams();
            ParamStrings.Add(cmd);
            if (cwd != null) {
                ParamStrings.Add(cwd);    
            }

            JToken result = SendRequest();
            if (result == null) {
                Console.WriteLine("Nope nope nope");
                return null;
            }

            return result.Value<string>();
        }
    }
}

File: src/natives/python/tool_request.py
import request_generator


class ToolRequest(request_generator.RequestGenerator):
    def __init__(self, client):
        super(ToolRequest, self).__init__(client)
        self.type = 'action'
        self.device = 'tool'

    def get_clipboard(self):
        self.action = 'get_clipboard'
        self.params = []
        return self.send_request()

    def set_clipboard(self, data):
        self.action = 'set_clipboard'
        self.params = [data]
        return self.send_request()

    def execute(self, cmd, cwd = None):
        self.action = 'execute'
        self.params = [cmd]
        if cwd is not None:
            self.params.append(cwd)

        return self.send_request()
